kep/205-preemption-toleration/README.md

```diff
@@ -0,0 +1,126 @@
+
+So, cluster administrators would like to exempt pods that do not finish the first epoch from preemption even if the job has lower priority. But, it is hard to monitor when the first epoch is finished outside of the job. Thus, a minimum running time is useful practically.
+
+For example, cluster administrator would introduce several priority classes with the same priority value for introducing several minimum running time guarantee categories like below:
+
+```yaml
+    system-critical: 10000
+                low:  8000  # no minimum running time guarantee
+          low-10min:  8000  # 10 min minimum running time guarantee against 8000 < p < 10000
+          low-30min:  8000  # 30 min minimum running time guarantee against 8000 < p < 10000
+```
+
+## Design Details
+
+### Preemption Toleration API
+
+Preemption toleration policy is defined in this struct type.
+
+```go
+// Please note that this does NOT have ObjectMeta 
+// because this will be embedded in annotation in PriorityClass as unnamed object.
+type PreemptionToleration struct {
+	metav1.TypeMeta `json:",inline"`
+
+  // MinimumPreemptablePriority specifies the minimum priority value that can preempt this priority class.
+  MinimumPreemptablePriority *int32 `json:"minimumPreemptablePriority,omitempty"`
+
+  // TolerationSeconds specified how many seconds this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.  Null value specifies forever (default).  Duration means the duration from the pod being scheduled to some node.  This value affects only to scheduled pods (no effect on nominated nodes).
+  TolerationSeconds *int64 `json:"tolerationSeconds,omitempty"`
+
+  // more policy will arise in the future.
+}
+```
+
+Users specifies the preemption toleration policy as an annotation with the key `scheduling.sigs.k8s.io/preemption-toleration` in `PriorityClass` as below:
+
+```yaml
+kind: PriorityClass
+metadata:
+  name: toleration-policy-sample
+  annotation:
+    scheduling.sigs.k8s.io/preemption-toleration: |
+      # the api is versioned for future enhancement
+      apiVersion: scheduling.sigs.k8s.io/v1beta1
+      kind: PreemptionToleration
+
+      # This priority class can tolerate preemption by priority with 8000 <= p < 10000.
+      minimumPreemptablePriority: 10000
+
+      # And it can tolerate preemption in 1 hour by the pod with priority (8000 <= p < 10000).
+      tolerationSeconds: 3600
+value: 8000
+```
+
+### Plugin implementation
+
+#### PostFilter
+
+This plugin implements the PostFilter extension point because this plugin focuses to extend the scheduler's preemption behavior.  Moreover, this plugin just extends victim candidate selection logics in the default preemption behavior. 
+
```

Comment:
I think you'd need to add a stanza to describe how to honor the preemption tolerationSeconds setting, e.g., implement the controller logic to monitor/manage the pods with preemption tolerationSeconds settings.

Reply:
Thanks. I will document where to how to modify the default preemption plugin.

Reply:
done in https://github.com/kubernetes-sigs/scheduler-plugins/pull/205/commits/8484681100265b185b73323495a71813f82bc715

Reply:
Ah, the details of PostFilter is awesome. But what I meant in the original comment was the "controller" logic - like how you want the controller to be implemented, launched and packaged (into scheduler or controller binary?).

(A reference is that Kubernetes uses a nodelifecycle manager to guarantee the tolerationSeconds' semantics)

Reply:
resolved in another thread https://github.com/kubernetes-sigs/scheduler-plugins/pull/205#issuecomment-883822936